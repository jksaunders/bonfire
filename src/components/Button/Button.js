import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Typography, { typographyPropTypes } from '../Typography';
import * as Colors from '../../theming/colors';
import { getVariant, withThemes } from '../../theming/ThemeContext';
import { css, cssBackground } from '../../styling';

const StyledButton = styled.button`
  ${props => cssBackground(props, { alpha: props.disabled ? 0.5 : 1 })}
  ${css('borderRadius', 'border-radius')}
  border-style: solid;
  border-width: 1px;
  ${css('padding')}
  width: ${({ width }) => width};
`;

const buttonStyleProps = {
  background: PropTypes.string,
  borderRadius: PropTypes.string,
  borderStyle: PropTypes.string,
  borderWidth: PropTypes.string,
  padding: PropTypes.string,
  typography: PropTypes.shape(typographyPropTypes),
};

const propTypes = {
  ...buttonStyleProps,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  text: PropTypes.string.isRequired,
  variant: PropTypes.oneOfType([
    PropTypes.shape(buttonStyleProps),
    PropTypes.string,
  ]),
  width: PropTypes.string,
};

const defaultProps = {
  /* eslint-disable react/default-props-match-prop-types */
  background: Colors.blue,
  borderRadius: null,
  borderStyle: null,
  borderWidth: null,
  padding: null,
  typography: null,
  /* eslint-enable react/default-props-match-prop-types */
  disabled: false,
  onClick: () => {},
  variant: null,
  width: '100px',
};

const Button = ({ disabled, onClick, text, ...props }) => {
  const { typography, ...buttonVariant } = getVariant(
    'button',
    props,
    buttonStyleProps
  );
  return (
    <StyledButton
      disabled={disabled}
      onClick={onClick}
      text={text}
      {...buttonVariant}
    >
      <Typography {...typography}>{text}</Typography>
    </StyledButton>
  );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

const ThemedButton = withThemes(Button);

export default ThemedButton;
