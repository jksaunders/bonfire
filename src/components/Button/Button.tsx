import React, { FC } from 'react';
import styled from 'styled-components';

const StyledButton = styled.button`
  background-color: ${(props): string => props.color};
`;

interface ButtonProps {
  color: string;
  children?: never;
}

const Button: FC<ButtonProps> = ({ color = 'red' }) => (
  <StyledButton color={color}>test button</StyledButton>
);

export default Button;
