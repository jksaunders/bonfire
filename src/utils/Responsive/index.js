export { default as ResponsiveRoot } from './ResponsiveRoot';
export * from './ResponsiveRoot';
