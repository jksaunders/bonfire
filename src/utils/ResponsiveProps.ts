import { SizeProp } from './Sizes';

export type ResponsiveProp<T> = T | SizeProp<T>;
export type ResponsiveProps<T, ResponsiveKeys = keyof T> = {
  [K in keyof T]: K extends ResponsiveKeys
    ? T[K] extends undefined
      ? ResponsiveProp<T[K]>
      : ResponsiveProp<NonNullable<T[K]>>
    : T[K];
};
