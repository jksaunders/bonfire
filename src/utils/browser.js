const goTo = (url) => {
  if (url && typeof window !== 'undefined') {
    window.location.href = url;
  }
};

export default {
  goTo
};