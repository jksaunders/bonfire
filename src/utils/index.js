export { default as Browser } from "./browser";
export * from "./styling";