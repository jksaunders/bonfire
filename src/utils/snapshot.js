/* eslint-disable import/no-extraneous-dependencies */
import { render as importedRender } from '@testing-library/react';
import 'jest-styled-components';

export const render = component => {
  const rendered = importedRender(component);
  return {
    ...rendered,
    component: rendered.container.firstChild,
  };
};

export const expectSnapshot = component => {
  expect(component).toMatchSnapshot();
};

export const expectExists = component => {
  expect(component).not.toEqual(null);
  expect(component).toBeInTheDocument();
};

export const expectStyle = (component, key, value) => {
  expect(component).toHaveStyleRule(key, value);
};
