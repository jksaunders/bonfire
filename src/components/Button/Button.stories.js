import React from "react";
import { ThemeProvider } from "styled-components";

import { action } from "@storybook/addon-actions";
import { select, withKnobs } from "@storybook/addon-knobs";

import { ThemeConstants } from "../../theming";
import Button from "./Button";

export default {
  title: "Components|Button",
  component: Button,
  decorators: [withKnobs]
};

export const ButtonStory = () => {
  const themeMode = select("Theme Mode", Object.keys(ThemeConstants.mode.values), Object.keys(ThemeConstants.mode.values)[0]);
  const themeLayout = select("Theme Layout", Object.keys(ThemeConstants.layout.values), Object.keys(ThemeConstants.layout.values)[0]);
  const theme = { mode: themeMode, layout: themeLayout };

  return (
    <ThemeProvider theme={theme}>
      <div>
        <div>
          {Object.keys(Button.CONSTANTS.VARIANT).map(variantKey => (
            <Button key={variantKey} text={variantKey.toLowerCase()} variant={variantKey} onClick={action("clicked")} />
          ))}
        </div>
        <div>
          {Object.keys(Button.CONSTANTS.VARIANT).map(variantKey => (
            <Button key={variantKey} text={`${variantKey.toLowerCase()} disabled`} variant={variantKey} onClick={action("clicked")} />
          ))}
        </div>
      </div>
    </ThemeProvider>
  );
};

ButtonStory.story = {
  name: "Button"
};