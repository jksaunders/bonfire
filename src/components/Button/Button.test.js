import React from "react";
import { mount } from "enzyme";
import { expectSnapshot } from "../../utils/snapshot";
import Button from "./Button";

const getButton = (props = {}) => <Button {...props} />;

test("Button mounts", () => {
  const button = getButton({ text: "Test text" });
  const wrapper = mount(button);
  expect(wrapper).toExist();
  expectSnapshot(button);
});