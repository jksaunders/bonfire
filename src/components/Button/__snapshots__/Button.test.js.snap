// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
.c1 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  font-weight: 500;
  color: #FFFFFF;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  -webkit-letter-spacing: 0.02857em;
  -moz-letter-spacing: 0.02857em;
  -ms-letter-spacing: 0.02857em;
  letter-spacing: 0.02857em;
  line-height: 1.75;
  font-size: 0.875rem;
  text-transform: uppercase;
}

.c0 {
  border-radius: 0.25em;
  border-style: solid;
  border-width: 0px;
  width: 100px;
}

<button
  class="c0"
  width="100px"
>
  <span
    class="c1"
    color="#FFFFFF"
    letter-spacing="0.02857em"
    transform="uppercase"
  >
    Test text
  </span>
</button>
`;
